src/dftbp/timedep/timeprop.F90

```diff
@@ -2759,9 +2886,93 @@ subroutine updateH0S(this, Ssqr, Sinv, coord, orb, neighbourList, nNeighbourSK,
 
     end if
 
+    call updateDQ(this, ints, neighbourList%iNeighbour, nNeighbourSK, img2CentCell, iSquare,&
+        & iSparseStart, Dsqr, Qsqr)
   end subroutine updateH0S
 
 
+  subroutine updateDQ(this, ints, iNeighbour, nNeighbourSK, img2CentCell, iSquare,&
+      & iSparseStart, Dsqr, Qsqr)
+
+    !> ElecDynamics instance
+    type(TElecDynamics), intent(inout) :: this
+
+    !> Integral container
+    type(TIntegral), intent(inout) :: ints
+
+    !> Atomic neighbour data
+    integer, intent(in) :: iNeighbour(0:,:)
+
+    !> Number of neighbours for each of the atoms
+    integer, intent(in) :: nNeighbourSK(:)
+
+    !> Index array for start of atomic block in dense matrices
+    integer, intent(in) :: iSquare(:)
+
+    !> index array for location of atomic blocks in large sparse arrays
+    integer, intent(in) :: iSparseStart(0:,:)
+
+    !> image atoms to their equivalent in the central cell
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Square dipole matrix
+    complex(dp), intent(inout), optional :: Dsqr(:,:,:,:)
+
+    !> Square quadrupole matrix
+    complex(dp), intent(inout), optional :: Qsqr(:,:,:,:)
+
+    real(dp), allocatable :: M3(:, :, :)
+    integer :: iSpin, iKS, iK
+
+    if (present(Dsqr)) then
+      Dsqr(:,:,:,:) = 0.0_dp
+      if (this%tRealHS) then
+        if (allocated(M3)) deallocate(M3)
+        allocate(M3(this%nDipole, this%nOrbs, this%nOrbs))
+        do iKS = 1, this%parallelKS%nLocalKS
+          iK = this%parallelKS%localKS(1, iKS)
+          iSpin = this%parallelKS%localKS(2, iKS)
+          call unpackDQ(M3, ints%dipoleBra, ints%dipoleKet, iNeighbour,&
+              & nNeighbourSK, iSquare, iSparseStart, img2CentCell)
+          Dsqr(:,:,:,iKS) = cmplx(M3, 0, dp)
+        end do
+      else
+        do iKS = 1, this%parallelKS%nLocalKS
+          iK = this%parallelKS%localKS(1, iKS)
+          iSpin = this%parallelKS%localKS(2, iKS)
+          call unpackDQ(Dsqr(:,:,:,iKS), ints%dipoleBra, ints%dipoleKet,&
+              & this%kPoint(:,iK), iNeighbour, nNeighbourSK, this%iCellVec, this%cellVec,&
+              & iSquare, iSparseStart, img2CentCell)
+        end do
+      end if
+    end if
+
+    if (present(Qsqr)) then
+      Qsqr(:,:,:,:) = 0.0_dp
+      if (this%tRealHS) then
+        if (allocated(M3)) deallocate(M3)
```

Comment:
The same here.

Reply:
Here it can be still allocated from unpacking the dipole integrals, which uses the same structure due to symmetry ;)